src/java.base/share/classes/java/security/PEMRecord.java

```diff
@@ -0,0 +1,157 @@
+ *
+ * @param type The type identifier in the PEM header.  For a public key,
+ * {@code type} would be "PUBLIC KEY".
+ * @param pem Any data between the PEM header and footer.
+ * @param leadingData Any non-PEM data read during the decoding process
+ * before the PEM header. This can be useful when reading metadata that
+ * accompanies PEM data.
+ *
+ * @spec https://www.rfc-editor.org/info/rfc7468
+ *       RFC 7468: Textual Encodings of PKIX, PKCS, and CMS Structures
+ *
+ * @see PEMDecoder
+ * @see PEMEncoder
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.PEM_API)
+public record PEMRecord(String type, String pem, byte[] leadingData)
+    implements DEREncodable {
+
+    /**
+     * Creates a {@code PEMRecord} instance with the given parameters.
+     *
+     * @param type the type identifier in the PEM header and footer, or
+     *             {@code null} if there is no PEM data.
+     * @param pem the data between the PEM header and footer.
+     * @param leadingData Any non-PEM data read during the decoding process
+     *                    before the PEM header.  This value maybe {@code null}.
+     * @throws IllegalArgumentException on incorrect input values
+     */
+    public PEMRecord(String type, String pem, byte[] leadingData) {
+        if (type == null && pem == null && leadingData == null) {
+            throw new IllegalArgumentException("All values may not be null.");
+        }
+
+        if (type == null && pem != null || type != null && pem == null) {
+            throw new IllegalArgumentException("\"type\" and \"pem\" must be" +
+                " both null or non-null");
+        }
+
+        // With no validity checking on `type`, the constructor accept anything
+        // including lowercase.  The onus is on the caller.
+        if (type != null && (type.startsWith("-") || type.contains("BEGIN") ||
+            type.contains("END") || type.endsWith("-"))) {
+            throw new IllegalArgumentException("Only the PEM type identifier " +
+                "is allowed");
+        }
+
+        this.type = type;
+        this.pem = pem;
+        this.leadingData = leadingData;
+    }
+
+    /**
+     * Creates a {@code PEMRecord} instance with a given {@code type} and {@code pem}
+     * data in String form.  {@code leadingData} is set to null.
+     *
+     * @param type the type identifier in the PEM header and footer, or {@code null} if there is no PEM data.
+     * @param pem The data between the PEM header and footer.
+     *
+     * @see #PEMRecord(String, String, byte[])
+     */
+    public PEMRecord(String type, String pem) {
+        this(type, pem, null);
+    }
+
+    /**
+     * Creates a {@code PEMRecord} instance with a given String {@code type} and
+     * byte array {@code pem}.  {@code leadingData} is set to null.
+     *
+     * @param type the type identifier in the PEM header and footer, or {@code null} if there is no PEM data.
+     * @param pem the data between the PEM header and footer.
```

Comment:
I thought the bytes were the DER encoded data. Wouldn't that be useful, if you were trying to PEM encode a certificate with extra leading data, for example?